logger/transport.go

```diff
@@ -0,0 +1,139 @@
+package logger
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+)
+
+type transport struct {
+	key       string
+	buffer    []Message
+	options   Options
+	closeChan chan bool
+	bufMu     sync.Mutex
+}
+
+func newTransport(options Options, key string) *transport {
+	return &transport{
+		key:       key,
+		options:   options,
+		closeChan: make(chan bool),
+	}
+}
+
+func (t *transport) start() {
+	ticker := time.NewTicker(t.options.FlushInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			t.drain()
+		case <-t.closeChan:
+			return
+		}
+	}
+}
+
+func (t *transport) close() {
+	if len(t.buffer) > 0 {
+		t.drain()
+	}
+
+	t.closeChan <- true
+}
+
+func (t *transport) add(msg Message) {
+	t.bufMu.Lock()
+	t.buffer = append(t.buffer, msg)
+	t.bufMu.Unlock()
+
+	// drain if the max buffer length is hit
+	if len(t.buffer) >= t.options.MaxBufferLen {
+		t.drain()
+	}
+}
+
+func (t *transport) drain() {
+	t.bufMu.Lock()
+	defer t.bufMu.Unlock()
+
+	size := len(t.buffer)
+	if size == 0 {
+		return
+	}
+
+	// TODO(mdeltito): in the future a retry should be triggered,
+	// and the buffer should only be cleared when the retry
+	// attempts are exhausted
+	msgs := t.buffer[:size]
+	t.send(msgs)
+	t.buffer = t.buffer[:0]
+}
+
+func (t *transport) send(msgs []Message) error {
+	var lines []Line
+	for _, msg := range msgs {
+		line := Line{
+			Body:  msg.Body,
+			App:   msg.Options.App,
+			Env:   msg.Options.Env,
+			Level: msg.Options.Level,
+		}
+
+		if msg.Options.Meta != "" {
+			line.Meta = metaEnvelope{
+				indexed: msg.Options.IndexMeta,
+				meta:    msg.Options.Meta,
+			}
+		}
+
+		lines = append(lines, line)
+	}
+
+	payload := Payload{
+		APIKey:     t.key,
+		Hostname:   t.options.Hostname,
+		IPAddress:  t.options.IPAddress,
+		MacAddress: t.options.MacAddress,
+		Tags:       t.options.Tags,
+		Lines:      lines,
+	}
+
+	pbytes, err := json.Marshal(payload)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest("POST", t.options.IngestURL, bytes.NewBuffer(pbytes))
+	req.Header.Set("user-agent", os.Getenv("USERAGENT"))
+	req.Header.Set("apikey", t.key)
+	req.Header.Set("Content-type", "application/json")
+
+	client := &http.Client{Timeout: t.options.SendTimeout}
+	resp, err := client.Do(req)
+
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= 500 {
+		return fmt.Errorf("Server error: %d", resp.StatusCode)
+	}
+
+	// TODO(mdeltito): do we need to parse the response?
+	// we arent doing anything with it
+	var apiresp ingestAPIResponse
+	err = json.NewDecoder(resp.Body).Decode(&apiresp)
+	if err != nil {
+		return err
+	}
```

Comment:
@dm36 this is more of an open question - the response code from ingestion is the only info we'll need to determine if we should retry, so it may not be necessary to parse the response

Reply:
hey yeah- I think we only really need the response code for the retry mechanism- but parsing the response has been helpful for debugging purposes

Reply:
leaving this comment in here for now, we can revisit as part of the remaining work and likely remove the parsing